src/generic_core/uproxy_core.spec.ts

```diff
@@ -110,3 +110,48 @@ describe('Core', () => {
     }).then(done);
   });
 });
+
+describe('retry', () => {
```

Comment:
Yay tests! Let's file an issue to extract `retry` and make it re-useable. I could imagine a `promises.ts` in uproxy-lib containing this and a couple others (alternatively, there are a bunch of libraries in NPM with this kind of thing already).